cpp/src/fil/infer.cu

```diff
@@ -845,21 +849,34 @@ __global__ void infer_k(storage_type forest, predict_params params)
   }
 }
 
-template <int NITEMS, leaf_algo_t leaf_algo>
+template <int NITEMS, typename real_t, leaf_algo_t leaf_algo>
 size_t shmem_size_params::get_smem_footprint()
 {
-  size_t finalize_footprint = tree_aggregator_t<NITEMS, leaf_algo>::smem_finalize_footprint(
+  size_t finalize_footprint = tree_aggregator_t<NITEMS, real_t, leaf_algo>::smem_finalize_footprint(
     cols_shmem_size(), num_classes, log2_threads_per_tree, predict_proba);
   size_t accumulate_footprint =
-    tree_aggregator_t<NITEMS, leaf_algo>::smem_accumulate_footprint(num_classes) +
+    tree_aggregator_t<NITEMS, real_t, leaf_algo>::smem_accumulate_footprint(num_classes) +
     cols_shmem_size();
   return std::max(accumulate_footprint, finalize_footprint);
 }
 
 template <class KernelParams>
 int compute_smem_footprint::run(predict_params ssp)
 {
-  return ssp.template get_smem_footprint<KernelParams::N_ITEMS, KernelParams::LEAF_ALGO>();
+  switch (ssp.sizeof_real) {
+    case 4:
+      return ssp
+        .template get_smem_footprint<KernelParams::N_ITEMS, float, KernelParams::LEAF_ALGO>();
+    case 8:
+      return ssp
+        .template get_smem_footprint<KernelParams::N_ITEMS, double, KernelParams::LEAF_ALGO>();
+    default:
+      ASSERT(ssp.sizeof_real == 4 || ssp.sizeof_real == 8,
```

Comment:
Rather than asserting a condition already checked by the switch, let's just do the necessary error handling. If we later add additional cases, we'll have to update 3 places with the current structure.

Reply:
Done.